net: websocket: pass request context to callback

mrodgers-witekio opened the pull request:
When upgrading an HTTP1 connection to a websocket, pass the HTTP request context associated with the upgrade request to the user callback.

This allows the user callback to access URL query parameters and HTTP headers. The main motivation behind this is to allow for authentication via the headers before the application decides whether to accept or reject the websocket connection, but it could also be used to provide any other context needed on opening the connection (e.g. `Sec-WebSocket-Protocol` header).

mrodgers-witekio committed:
net: websocket: pass HTTP upgrade request context to user callback

Passing HTTP upgrade request context to the user callback allows the
user to decide whether to accept or reject the websocket connection
based on the HTTP headers in the request. The primary reason for this is
to enable authentication of the websocket connection (e.g. via cookies
or Authorization header).

Signed-off-by: Matt Rodgers <mrodgers@witekio.com>

mrodgers-witekio committed:
doc: websocket: update docs to describe additional callback parameter

Update HTTP server documentation and migration guide to account for
added request_ctx parameter to the http_resource_websocket_cb_t
callback.

Signed-off-by: Matt Rodgers <mrodgers@witekio.com>

rlubos reviewed (dismissed):
It seems `shell_websocket.h` needs an update too (`shell_websocket_setup` declaration), looks good otherwise, thanks.

mrodgers-witekio commented:
> It seems `shell_websocket.h` needs an update too (`shell_websocket_setup` declaration), looks good otherwise, thanks.

Good spot, updated this and also removed the `extern` from the function declarations in the header file (don't think this is needed unless I missed something?)

rlubos commented:
> removed the extern from the function declarations in the header file (don't think this is needed unless I missed something?)

I don'tt think they're needed either, probably just some copy-paste leftover.
